File: android-app/app/src/main/java/com/lasohealth/android/feature/home/PersonalHealthForecastCard.kt
package com.lasohealth.android.feature.home

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.NorthEast
import androidx.compose.material.icons.filled.East
import androidx.compose.material.icons.filled.SouthEast
import androidx.compose.material.icons.filled.ShowChart
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.MonitorHeart
import androidx.compose.material.icons.filled.Bedtime
import androidx.compose.material.icons.filled.DirectionsRun
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.lasohealth.android.core.model.ForecastDirection
import com.lasohealth.android.core.model.HealthMetric
import com.lasohealth.android.core.model.MetricForecastUi
import com.lasohealth.android.ui.theme.AccentGreen
import com.lasohealth.android.ui.theme.AccentOrange

/**
 * Displays forward-looking metric forecasts with statistically guaranteed confidence intervals.
 * Shows predicted values, direction arrows, and confidence percentages.
 * iOS parity: PersonalHealthForecastCard.swift
 */
@Composable
fun PersonalHealthForecastCard(
    forecasts: List<MetricForecastUi>,
    onTapMetric: (HealthMetric) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (forecasts.isEmpty()) return

    Surface(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp),
        shape = RoundedCornerShape(16.dp),
        color = MaterialTheme.colorScheme.surface,
        border = BorderStroke(
            width = 0.5.dp,
            color = MaterialTheme.colorScheme.outline.copy(alpha = 0.2f),
        ),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            // Header
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Icon(
                        imageVector = Icons.Filled.ShowChart,
                        contentDescription = null,
                        modifier = Modifier.size(14.dp),
                        tint = Color(0xFF3478F6),
                    )
                    Text(
                        text = "YOUR FORECAST",
                        style = MaterialTheme.typography.labelSmall,
                        fontWeight = FontWeight.SemiBold,
                        letterSpacing = 1.sp,
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
                    )
                }
                Text(
                    text = "Next 7 days",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.35f),
                )
            }

            // Forecast rows
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                forecasts.take(3).forEach { forecast ->
                    ForecastRow(
                        forecast = forecast,
                        onClick = { onTapMetric(forecast.metric) },
                    )
                }
            }
        }
    }
}

@Composable
private fun ForecastRow(
    forecast: MetricForecastUi,
    onClick: () -> Unit,
) {
    val directionColor = when {
        forecast.directionIcon == ForecastDirection.STABLE -> MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f)
        forecast.isFavorable -> AccentGreen
        else -> AccentOrange
    }

    val directionVector: ImageVector = when (forecast.directionIcon) {
        ForecastDirection.UP -> Icons.Filled.NorthEast
        ForecastDirection.DOWN -> Icons.Filled.SouthEast
        ForecastDirection.STABLE -> Icons.Filled.East
    }

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(vertical = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        // Metric icon badge
        Box(
            modifier = Modifier
                .size(28.dp)
                .background(
                    color = directionColor.copy(alpha = 0.12f),
                    shape = RoundedCornerShape(7.dp),
                ),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = metricIcon(forecast.metric),
                contentDescription = null,
                modifier = Modifier.size(14.dp),
                tint = directionColor,
            )
        }

        // Metric name + range
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Text(
                text = forecast.metric.displayName,
                style = MaterialTheme.typography.bodySmall,
                fontWeight = FontWeight.Medium,
                color = MaterialTheme.colorScheme.onSurface,
            )
            Text(
                text = forecast.rangeDescription,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
            )
        }

        // Direction + value + confidence
        Column(
            horizontalAlignment = Alignment.End,
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(3.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    imageVector = directionVector,
                    contentDescription = null,
                    modifier = Modifier.size(12.dp),
                    tint = directionColor,
                )
                Text(
                    text = forecast.predictedValueFormatted,
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.SemiBold,
                    color = directionColor,
                )
            }
            Text(
                text = "${forecast.confidencePercent}% conf",
                style = MaterialTheme.typography.labelSmall.copy(fontSize = 9.sp),
                fontWeight = FontWeight.Medium,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
            )
        }
    }
}

/** Map metric to a Material icon for forecast rows. */
@Composable
private fun metricIcon(metric: HealthMetric): ImageVector {
    return when (metric) {
        HealthMetric.HEART_RATE, HealthMetric.RESTING_HEART_RATE -> Icons.Filled.Favorite
        HealthMetric.HEART_RATE_VARIABILITY -> Icons.Filled.MonitorHeart
        HealthMetric.SLEEP_DURATION, HealthMetric.SLEEP_DEEP, HealthMetric.SLEEP_REM -> Icons.Filled.Bedtime
        HealthMetric.STEPS, HealthMetric.ACTIVE_CALORIES -> Icons.Filled.DirectionsRun
        HealthMetric.VO2_MAX -> Icons.Filled.ShowChart
        else -> Icons.Filled.ShowChart
    }
}

File: Views/Home/PersonalHealthForecastCard.swift
import SwiftUI

// MARK: - Personal Health Forecast Card

/// Displays forward-looking metric forecasts with statistically guaranteed confidence intervals.
/// Based on conformal prediction research (Communications Engineering 2025): wraps predictions
/// in intervals with guaranteed coverage, creating a "predicted you" vs actual tracking.
struct PersonalHealthForecastCard: View {
    let forecasts: [MetricForecast]
    let onTapMetric: (HealthMetric) -> Void

    var body: some View {
        if !forecasts.isEmpty {
            VStack(alignment: .leading, spacing: 12) {
                // Header
                HStack(spacing: 6) {
                    Image(systemName: "chart.line.uptrend.xyaxis")
                        .font(.caption2.weight(.bold))
                        .foregroundStyle(.blue)
                    Text("YOUR FORECAST")
                        .font(.caption.weight(.semibold))
                        .foregroundStyle(.secondary)
                        .textCase(.uppercase)
                    Spacer()
                    Text("Next 7 days")
                        .font(.caption2)
                        .foregroundStyle(.tertiary)
                }

                // Forecast rows
                VStack(spacing: 8) {
                    ForEach(Array(forecasts.prefix(3))) { forecast in
                        forecastRow(forecast)
                            .onTapGesture {
                                AppAnalytics.shared.trackBlockTap(title: "Forecast Metric", type: .metricRow, screen: .home, metadata: ["source": "health_forecast", "metric": forecast.metric.rawValue])
                                onTapMetric(forecast.metric)
                            }
                    }
                }
            }
            .padding(DS.space4)
            .background(.background, in: RoundedRectangle(cornerRadius: 16))
            .overlay(
                RoundedRectangle(cornerRadius: 16)
                    .strokeBorder(.quaternary, lineWidth: 0.5)
            )
            .padding(.horizontal)
        }
    }

    // MARK: - Forecast Row

    private func forecastRow(_ forecast: MetricForecast) -> some View {
        HStack(spacing: 12) {
            // Metric icon
            Image(systemName: forecast.metric.systemImageName)
                .font(.caption)
                .foregroundStyle(forecast.directionColor)
                .frame(width: 28, height: 28)
                .background(forecast.directionColor.opacity(0.12), in: RoundedRectangle(cornerRadius: 7))

            // Info
            VStack(alignment: .leading, spacing: 2) {
                Text(forecast.metric.displayName)
                    .font(.caption.weight(.medium))
                    .foregroundStyle(.primary)

                Text(forecast.rangeDescription)
                    .font(.caption2)
                    .foregroundStyle(.secondary)
            }

            Spacer()

            // Direction indicator
            VStack(alignment: .trailing, spacing: 2) {
                HStack(spacing: 3) {
                    Image(systemName: forecast.directionIcon)
                        .font(.caption2)
                    Text(forecast.predictedValueFormatted)
                        .font(.caption.weight(.semibold))
                }
                .foregroundStyle(forecast.directionColor)

                // Confidence badge
                Text("\(forecast.confidencePercent)% conf")
                    .font(.system(size: 9, weight: .medium))
                    .foregroundStyle(.secondary)
            }
        }
        .padding(.vertical, DS.space1)
    }
}

// MARK: - Metric Forecast Model

struct MetricForecast: Identifiable {
    let id = UUID()
    let metric: HealthMetric
    let predictedValue: Double
    let lowerBound: Double
    let upperBound: Double
    let currentValue: Double
    let horizon: Int  // days ahead
    let confidence: Double  // 0-1

    /// Percentage confidence for display
    var confidencePercent: Int { Int(confidence * 100) }

    /// Direction of predicted change
    var direction: Direction {
        let change = predictedValue - currentValue
        let threshold = currentValue * 0.03  // 3% threshold
        if abs(change) < threshold { return .stable }
        return change > 0 ? .up : .down
    }

    /// Whether the predicted direction is favorable
    var isFavorable: Bool {
        switch direction {
        case .up: return metric.higherIsBetter
        case .down: return !metric.higherIsBetter
        case .stable: return true
        }
    }

    var directionIcon: String {
        switch direction {
        case .up: return "arrow.up.right"
        case .down: return "arrow.down.right"
        case .stable: return "arrow.right"
        }
    }

    var directionColor: Color {
        if direction == .stable { return .secondary }
        return isFavorable ? .green : .orange
    }

    var predictedValueFormatted: String {
        formatMetricValue(predictedValue, metric: metric)
    }

    var rangeDescription: String {
        let lower = formatMetricValue(lowerBound, metric: metric)
        let upper = formatMetricValue(upperBound, metric: metric)
        return "Expected: \(lower) – \(upper)"
    }

    enum Direction { case up, down, stable }

    private func formatMetricValue(_ value: Double, metric: HealthMetric) -> String {
        switch metric {
        case .steps:
            return "\(Int(value))"
        case .sleepDuration, .sleepDeep, .sleepREM, .sleepCore:
            let h = value / 3600
            return String(format: "%.1fh", h)
        case .heartRateVariability:
            return String(format: "%.0fms", value)
        case .heartRate, .restingHeartRate:
            return String(format: "%.0f bpm", value)
        case .vo2Max:
            return String(format: "%.1f", value)
        case .weight:
            return String(format: "%.1f", value)
        default:
            if value >= 1000 { return String(format: "%.0f", value) }
            if value >= 10 { return String(format: "%.1f", value) }
            return String(format: "%.2f", value)
        }
    }
}

// MARK: - Forecast Builder (bridges MLOrchestrator → View)

enum ForecastBuilder {

    /// Build MetricForecast array from MLOrchestrator's multi-horizon forecasts
    static func buildForecasts(
        multiHorizonForecasts: [HealthMetric: TimeSeriesForecaster.MultiHorizonForecast],
        timeSeries: [HealthMetric: MetricTimeSeries],
        maxCount: Int = 3
    ) -> [MetricForecast] {
        var forecasts: [MetricForecast] = []

        // Priority metrics to show forecasts for
        let priorityMetrics: [HealthMetric] = [
            .heartRateVariability, .restingHeartRate, .sleepDuration,
            .steps, .vo2Max, .activeCalories, .sleepDeep
        ]

        for metric in priorityMetrics {
            guard let multiHorizon = multiHorizonForecasts[metric],
                  let series = timeSeries[metric],
                  let currentValue = series.samples.last?.value,
                  let horizon7 = multiHorizon.horizons.first(where: { $0.horizon == 7 })
                    ?? multiHorizon.horizons.last else {
                continue
            }

            let ciWidth = horizon7.ciUpper - horizon7.ciLower
            let confidence = max(0.5, min(0.95, 1.0 - ciWidth / max(1, abs(horizon7.value) * 2)))

            forecasts.append(MetricForecast(
                metric: metric,
                predictedValue: horizon7.value,
                lowerBound: horizon7.ciLower,
                upperBound: horizon7.ciUpper,
                currentValue: currentValue,
                horizon: horizon7.horizon,
                confidence: confidence
            ))

            if forecasts.count >= maxCount { break }
        }

        return forecasts
    }
}

#Preview {
    let forecasts: [MetricForecast] = [
        MetricForecast(
            metric: .heartRateVariability,
            predictedValue: 52,
            lowerBound: 42,
            upperBound: 62,
            currentValue: 48,
            horizon: 7,
            confidence: 0.85
        ),
        MetricForecast(
            metric: .sleepDuration,
            predictedValue: 27000,
            lowerBound: 24000,
            upperBound: 30000,
            currentValue: 25200,
            horizon: 7,
            confidence: 0.78
        ),
        MetricForecast(
            metric: .restingHeartRate,
            predictedValue: 58,
            lowerBound: 55,
            upperBound: 61,
            currentValue: 60,
            horizon: 7,
            confidence: 0.82
        ),
    ]
    VStack {
        PersonalHealthForecastCard(
            forecasts: forecasts,
            onTapMetric: { _ in }
        )
    }
    .padding(.vertical)
    .background(Color(.systemGroupedBackground))
}
